src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                            parameterIsNamed, cancellationToken);
+                    }
+
+                    var expressionFromInvocation = syntaxFacts.GetExpressionOfInvocationExpression(invocationExpression);
+                    var allArguments = AddArgumentToArgumentList(updatedInvocationArguments, generator,
+                        newArgumentExpression.WithAdditionalAnnotations(Formatter.Annotation), insertionIndex, parameterName, parameterIsNamed);
+                    var newInvo = editor.Generator.InvocationExpression(expressionFromInvocation, allArguments);
+                    return newInvo;
+                });
+            }
+
+            // If you are at the original document, then also introduce the new method and introduce the parameter.
+            if (document.Id == originalDocument.Id)
+            {
+                await UpdateExpressionInOriginalFunctionAsync(originalDocument, expressionCopy, containingMethod,
+                    parameterName, editor, allOccurrences, cancellationToken).ConfigureAwait(false);
+                var semanticModel = await originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var parameterType = semanticModel.GetTypeInfo(expressionCopy, cancellationToken).Type ??
+                    semanticModel.Compilation.ObjectType;
+                var refKind = syntaxFacts.GetRefKindOfArgument(expressionCopy);
+                var parameter = generator.ParameterDeclaration(name: parameterName,
+                    type: generator.TypeExpression(parameterType), refKind: refKind);
+                editor.InsertParameter(containingMethod, insertionIndex, parameter);
+            }
+
+            return modifiedSolution.WithDocumentSyntaxRoot(originalDocument.Id, editor.GetChangedRoot());
+        }
+
+        /// <summary>
+        /// Goes through all of the invocations and replaces the expression with identifiers from the invocation 
+        /// arguments and rewrites the call site with the updated expression as a new argument.
+        /// </summary>
+        private async Task<Solution> RewriteSolutionAsync(Document originalDocument,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            ImmutableDictionary<Document, ImmutableArray<TInvocationExpressionSyntax>> callSites,
+            bool allOccurrences, string parameterName, CancellationToken cancellationToken)
+        {
+            // Need a copy of the expression to use to find the original expression
+            // because it gets modified when tracking the nodes.
+            var expressionCopy = expression;
+            var mappingDictionary = await MapExpressionToParametersAsync(originalDocument, expression, cancellationToken).ConfigureAwait(false);
+
+            // Need to track the nodes in the expression so that they can be found later.
```

Comment:
i don't undersatand this.  either this is unneecssary, or this comment isn't sufficient at all as to explaining why we need this.

Reply:
removed tracking